pkg/astvalidation/rule_known_type_names.go

```diff
@@ -0,0 +1,116 @@
+package astvalidation
+
+import (
+	"github.com/cespare/xxhash"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+	"github.com/jensneuse/graphql-go-tools/pkg/operationreport"
+)
+
+func KnownTypeNames() Rule {
+	return func(walker *astvisitor.Walker) {
+		visitor := &knownTypeNamesVisitor{
+			Walker: walker,
+		}
+
+		walker.RegisterDocumentVisitor(visitor)
+		walker.RegisterEnterRootOperationTypeDefinitionVisitor(visitor)
+		walker.RegisterEnterFieldDefinitionVisitor(visitor)
+		walker.RegisterEnterUnionMemberTypeVisitor(visitor)
+		walker.RegisterEnterInputValueDefinitionVisitor(visitor)
+		walker.RegisterEnterObjectTypeDefinitionVisitor(visitor)
+		walker.RegisterEnterInterfaceTypeDefinitionVisitor(visitor)
+		walker.RegisterEnterScalarTypeDefinitionVisitor(visitor)
+		walker.RegisterEnterUnionTypeDefinitionVisitor(visitor)
+		walker.RegisterEnterInputObjectTypeDefinitionVisitor(visitor)
+		walker.RegisterEnterEnumTypeDefinitionVisitor(visitor)
+	}
+}
+
+type knownTypeNamesVisitor struct {
+	*astvisitor.Walker
+	definition           *ast.Document
+	definedTypeNameHashs map[uint64]bool
+	referencedTypeNames  map[uint64][]byte
+}
+
+func (u *knownTypeNamesVisitor) EnterDocument(operation, definition *ast.Document) {
+	u.definition = definition
+	u.definedTypeNameHashs = make(map[uint64]bool)
+	u.referencedTypeNames = make(map[uint64][]byte)
+}
+
+func (u *knownTypeNamesVisitor) LeaveDocument(operation, definition *ast.Document) {
+	for referencedTypeNameHash, referencedTypeName := range u.referencedTypeNames {
+		if !u.definedTypeNameHashs[referencedTypeNameHash] {
+			u.StopWithExternalErr(operationreport.ErrTypeUndefined(referencedTypeName))
+			continue
```

Comment:
should we break after error?

or maybe it will be better to call add error directly?
```
w.Report.AddExternalError(err)
```

Reply:
your suggestion is way better